### PR TITLE
Document symex instructions

### PR DESCRIPTION
**Pull request by NlightNFotis**
Add documentation for the `goto-symex` and `goto-program-instructions`.

### CODE REVIEW

File: doc/architectural/central-data-structures.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+The following is some light technical documentation of the major data structures
+represented in the input transformation to Intermediate Representation (IR) inside
+CBMC and the assorted CProver tools.
+
+## goto_modelt
+
+The `goto_modelt` is the main data structure that CBMC (and the other tools) use to
+represent GOTO-IR (the `GOTO` Intermediate Representation).
+
+A `goto_modelt` is effectively a product type, consisting of:
```
Comment (martin-cs) [inline doc/architectural/central-data-structures.md]: Same comment as on the previous draft; I would suggest using "pair" or "tuple" as a more C++-ish way of saying this.

File: doc/architectural/central-data-structures.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+* A list of GOTO functions (pseudocode: `type goto_functionst = list<goto_functiont>`)
```
Comment (martin-cs) [inline doc/architectural/central-data-structures.md]: It's not really a list...

File: doc/architectural/central-data-structures.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+The abstract interface of `goto_modelt` is outlined in the file
+[`src/goto-programs/abstract_goto_model.h`](../../src/goto-programs/abstract_goto_model.h).
+
+## goto_functiont
+
+A `goto_functiont` is also a product type. It's designed to represent a function
+at the IR level, and effectively it's the following ADT (in pseudocode):
+
+```js
```
Comment (martin-cs) [inline doc/architectural/central-data-structures.md]: Could you use C++-style pseudocode?

Comment (martin-cs) [reply doc/architectural/central-data-structures.md]: And / or a reference to the code.

File: doc/architectural/central-data-structures.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+* The `guard` is an `exprt` (A CBMC data structure broadly representing an expression)
+  that is expected to have type `bool`.
+  * This is optional - not every instruction is expected to have a `guard` associated with it.
+
+## source_locationt
+
+Another important data structure that needs to be discussed at this point is
+`source_locationt`.
+
+This is an `irept`. `irep`s are the central data structure that model most entities inside
```
Comment (martin-cs) [inline doc/architectural/central-data-structures.md]: `irept` is pretty critical and deserves space of its own, not just under `source_locationt`

File: doc/architectural/symex-instructions.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+* An instruction guard (an expression that needs to be evaluated to `true` before
+  the instruction is to be executed - if one is attached to the instruction).
+
+In this document, we are going to take a closer look at the first subcomponent,
+the instruction types, along with the interplay between the instructions and a
+central CBMC module, the *symbolic execution engine* (from now on, just *symex*).
+
+## A (very) short introduction to Symex
+
+Symex is, at its core, a GOTO-program interpreter. While Symex is interpreting
```
Comment (martin-cs) [inline doc/architectural/symex-instructions.md]: 'an interpreter BUT using symbolic values, not literal ones. For example...  Note that this produces a formula which describes all possible outputs rather than a single output value.'

File: doc/architectural/symex-instructions.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+```c
+int main(int argc, char **argv)
+{
+    int arry[] = {0, 1, 2, 3};
+    __CPROVER_assert(arry[3] != 3, "expected failure: last arry element is equal to 3");
+}
+```
+
+If we ask CBMC to print the GOTO-program instructions with `--show-goto-functions`, then part
```
Comment (martin-cs) [inline doc/architectural/symex-instructions.md:72]: Should be in the `goto-functions` documentation rather than just in symex.

File: regression/cbmc/field-sensitivity15/test.desc [hunk 7]
```diff
@@ -0,0 +1,9 @@
+CORE
+main.c
+
+^VERIFICATION SUCCESSFUL$
+^EXIT=0$
+^SIGNAL=0$
+--
+--
+The test is a minimized version of code generated from s2n-quic using Kani.
```
Comment (martin-cs) [inline regression/cbmc/field-sensitivity15/test.desc]: More tests are always good but why is this in a documentation PR?  I guess I have missed something...

File: src/goto-symex/symex_assign.cpp [hunk 8]
```diff
@@ -215,14 +215,6 @@ void symex_assignt::assign_non_struct_symbol(
       assignment.rhs,
       target,
       symex_config.allow_pointer_unsoundness);
-    // Erase the composite symbol from our working state. Note that we need to
```
Comment (martin-cs) [inline src/goto-symex/symex_assign.cpp]: There could well be a good justification for this removal of code but please could they not be in a documentation PR?

**Review by martin-cs — CHANGES_REQUESTED** [review]

**Comment by NlightNFotis** [comment]
Hi @martin-cs , this was branched of at an earlier point from the same branch as the other documentation PR (#7470), and it hasn't tracked the changes that addressed the comments on the other PR, so it appears like it has the same problems, but it's just being stale.

This means that the earlier commits here are just adding noise, so I will focus on addressing everything on the other PR, get it merged, and then rebase this one on top of `develop`, for a cleaner, less stale history. I will then address the comments that remain relevant after the rebase and ask you for a re-review.

File: doc/architectural/cbmc-architecture.md [hunk 9]
```diff
@@ -64,9 +71,10 @@ To be documented.
 
 To be documented.
 
-## Goto functions &rarr; BMC &rarr; Counterexample (trace) ##
+## Goto functions &rarr; BMC &rarr; Counterexample (trace)
```
Comment (jimgrundy) [inline doc/architectural/cbmc-architecture.md:74]: Why this change? Doesn't seem to match the other lines.

File: doc/architectural/central-data-structures.md (hunk 4 above)
Comment (jimgrundy) [reply doc/architectural/central-data-structures.md]: +1

File: doc/architectural/symex-instructions.md [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+Symex's source code is available under [src/goto-symex](../../src/goto-symex/).
+
+## Instruction Types
+
+Every GOTO-program instruction has a specific type. You can see the instruction type
+at #goto_program_instruction_typet but we will also list some of them here for illustration
+purposes:
+
+```c
+enum goto_program_instruction_typet
```
Comment (jimgrundy) [inline doc/architectural/symex-instructions.md:40]: Maybe throw in some "..."s here to emphasize that this isn't the full set.

File: doc/architectural/symex-instructions.md [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+{
+  GOTO = 1,              // branch, possibly guarded
+  ASSUME = 2,            // non-failing guarded self loop
```
Comment (jimgrundy) [inline doc/architectural/symex-instructions.md]: This is perhaps not the worlds most obvious description of an assumption.

File: doc/architectural/symex-instructions.md [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+list of what the instructions look like at this point in time.)
+
+You can observe these instruction types in the user-friendly print of the goto-program that
+various CPROVER programs produce with the flag `--show-goto-functions`. For a live example,
+consider the following C file:
+
+```c
+int main(int argc, char **argv)
+{
+    int arry[] = {0, 1, 2, 3};
```
Comment (jimgrundy) [inline doc/architectural/symex-instructions.md]: It this point you may as well just say "array", or go the whole way and call it "a". "arry" offers neither brevity nor clarity.

File: doc/architectural/symex-instructions.md (hunk 6 above)
Comment (jimgrundy) [reply doc/architectural/symex-instructions.md:72]: I presume (hope) it is. Maybe here we should have a cross reference to that part of the goto-functions documentation.

File: doc/architectural/symex-instructions.md [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+case ASSUME:
+  symex_assume(state, instruction.condition());
+  break;
+```
+
+The way the [`symex` subfolder](../../src/goto-symex/) is structured, the different
+dispatching functions are usually in their own file, designated by the instruction's
+name. As an example, you can find the code for the function goto_symext::symex_goto
+in [symex_goto.cpp](../../src/goto-symex/symex_goto.cpp)
+
```
Comment (jimgrundy) [inline doc/architectural/symex-instructions.md:122]: A description of what symex produces, why, and what it means could use a longer explanation.

File: src/goto-checker/multi_path_symex_only_checker.cpp [hunk 14]
```diff
@@ -74,8 +74,8 @@ void multi_path_symex_only_checkert::generate_equation()
 {
   const auto symex_start = std::chrono::steady_clock::now();
```
Comment (jimgrundy) [inline src/goto-checker/multi_path_symex_only_checker.cpp]: This PR seems like it should be a couple of PRs about different things. This doesn't feel like it belongs in a documentation PR.

**Review by jimgrundy — CHANGES_REQUESTED** [review]

File: doc/architectural/cbmc-architecture.md (hunk 9 above)
Comment (NlightNFotis) [reply doc/architectural/cbmc-architecture.md:74]: I did this as a drive-by clean up. As far as I know headings in markdown are only prepended by a `##`.

Comment (NlightNFotis) [reply doc/architectural/cbmc-architecture.md:74]: On a second visit, it appears that [doxygen does support any number of `#`](https://doxygen.nl/manual/markdown.html) after the title, but that's invalid markdown.

If this change offends in any capacity however, it's easy to be reverted.

Comment (tautschnig) [reply doc/architectural/cbmc-architecture.md:74]: It seems this file has that issue multiple times. Can you please clean up all occurrences? (src/goto-instrument/README.md also suffers from this problem. Maybe do one commit that does the cleanup for all of these?)

File: doc/architectural/symex-instructions.md [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+In this document, we are going to take a closer look at the first subcomponent,
+the instruction types, along with the interplay between the instructions and a
+central CBMC module, the *symbolic execution engine* (from now on, just *symex*).
+
+## A (very) short introduction to Symex
+
+Symex is, at its core, a GOTO-program interpreter. While Symex is interpreting
+the program, it's also building a list of SSA steps that form part of the equation
+that is to be sent to the solver.
```
Comment (tautschnig) [inline doc/architectural/symex-instructions.md]: Would it be better to refer the reader to `src/goto-symex/README.md`?

File: doc/architectural/symex-instructions.md [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+  // 3 file /tmp/example.c line 5 function main
+  DEAD main::1::arry
+  // 4 file /tmp/example.c line 5 function main
+  SET RETURN VALUE side_effect statement="nondet" is_nondet_nullable="1"
+  // 5 file /tmp/example.c line 5 function main
+  END_FUNCTION
+```
+
+In the above excerpt, the capitalised words at the beginning each instruction are the
+correspondent instruction types (`DECL`, `ASSIGN`, etc).
```
Comment (tautschnig) [inline doc/architectural/symex-instructions.md:94]: All of the above really belongs to goto programs, not so much to symex, does it?

File: doc/architectural/cbmc-architecture.md (hunk 9 above)
Comment (jimgrundy) [reply doc/architectural/cbmc-architecture.md:74]: I mean fair enough. I was thinking maybe there should be a separate PR that does this clean up. Or, at least, maybe consistently clean them all up in this file as there are many line this (as Michael as noticed). You want to leave the file in a consistent state so that newcomers will follow the example.

Comment (esteffin) [reply doc/architectural/cbmc-architecture.md:74]: I will make trailing `#` consistent by removing them in the files touched by this PR.

File: doc/architectural/symex-instructions.md (hunk 5 above)
Comment (esteffin) [reply doc/architectural/symex-instructions.md]: Agreed. Changed

File: doc/architectural/symex-instructions.md (hunk 15 above)
Comment (esteffin) [reply doc/architectural/symex-instructions.md]: Good idea, added

File: doc/architectural/symex-instructions.md (hunk 11 above)
Comment (esteffin) [reply doc/architectural/symex-instructions.md]: The reason for this is that it's a copy from `goto_program_instruction_typet` where the comment of `ASSUME` states what was written.

I changed it to `ASSUME` to avoid confusion.

File: doc/architectural/symex-instructions.md (hunk 12 above)
Comment (esteffin) [reply doc/architectural/symex-instructions.md]: Changed to `a` and improved assertion comment

File: doc/architectural/symex-instructions.md (hunk 16 above)
Comment (esteffin) [reply doc/architectural/symex-instructions.md:94]: Yes, but in this file we are describing the **Instruction type** that is exactly what we are showing.
In my opinion it is useful to show how the goto-program looks like.

**Review by jimgrundy — APPROVED** [review]

**Comment by esteffin** [comment]
@martin-cs Can you please have a look and confirm your concerns have been addressed?

**Review by TGWDB — APPROVED** [review]

**Comment by TGWDB** [comment]
@martin-cs Can you please act on this soon? It's been a week since the last request for review and more than that since various fixes were pushed. Note that we plan to force merge this in about 24 hours unless there is some new specific concern. We can make further changes in follow up tickets if necessary, but we'd like to merge this and progress ASAP.
